### DISCUSSION TITLE
Is any configuration needed to enable ECC features for Ryzen Processors in v7.0.0?

### DISCUSSION THREAD

**Discussion by rowanmoul** [memtest86plus/memtest86plus#436] [Q&A] [answered]
I have a Ryzen 5950x (vermeer) that I have paired with 128GB of ECC memory for use as a home server. It looks like support for ECC polling was added for this processor in the latest version (#353), but I am not clear on if I need to do anything to enable this feature or if it will just auto-detect and turn it on.
Looking at the code in that PR, it seems like it will auto-detect, but how would this show up in the on-screen output? I want to be 100% sure that ECC and EDAC is working with my motherboard.

**Comment by debrouxl** [comment, marked as answer]
Yes, you need to manually enable EDAC setup polling, by setting `enable_ecc_polling` to `true` in `app/config.c`, and recompiling memtest86+. Unlike pretty much all other configuration items, there's no command-line parameter (yet) for setting this variable to `true`.

Auto-detection of EDAC performed by `system/imc/*.c` files - currently, only `system/imc/amd_zen.c` - is conditional on that variable being `true`. When the EDAC is usable, the code sets `ecc_status.ecc_enabled` to `true`, which in turn enables both subsequent EDAC polling and visual feedback for ECC error reporting enablement: on the right side of line 8, "ECC" and two error counts are displayed when enabled.

Currently, from `app/display.c` and `app/display.h`:
```
    prints(8, 0, "Using:                                    | Pass:           Errors:");
//  prints(9, 0, "--------------------------------------------------------------------------------");

    if (ecc_status.ecc_enabled) {
        prints(8, 57, "Err:        ECC:");
    }
```

```
void display_error_count(void)
{
    if (ecc_status.ecc_enabled) {
        display_err_count_with_ecc(error_count, error_count_cecc);
    } else {
        display_err_count_without_ecc(error_count);
    }
}
```

```
#define display_err_count_without_ecc(count) \
    printi(8, 68, count, 0, false, true)

#define display_err_count_with_ecc(count_err, count_ecc) \
    { \
        printi(8, 62, count_err, 0, false, true); \
        printi(8, 74, count_ecc, 0, false, true); \
    }
```
(copying the current code snippets instead of linking them, as line numbers can change over time)

**Reply by edo1** [reply]
I'm using a Ryzen 7700 with slightly overclocked ECC memory. PassMark's MemTest86 detects corrected ECC errors. The Linux kernel also reports corrected ECC errors when running memory-intensive applications like memtester and mprime.

I've set `enable_ecc_polling` to `true` in app/config.c and confirmed that `ecc_status.ecc_enabled` is `true`, so ECC detection logic seems active.


@x86fr, Is the ECC support code in Memtest86+ complete? It looks like it should work, but I haven't seen any ECC errors detected.

**Reply by x86fr** [reply]
For AMD Raphael, it should work as expected. Do you see the `ECC:` count on main screen? Can you please post a screenshot? 
Did you tried with latest nightly build? What is your motherboard model and how many DDR5 ECC stick do you have?

**Reply by edo1** [reply]
![image](https://github.com/user-attachments/assets/10acbf78-7c40-4613-9cf4-cf6b31397f41)
(full pass w/o errors was done, but I haven't taken a photo)

applied changed:
```
diff --git a/app/config.c b/app/config.c
index 83e58e8..8c0db68 100644
--- a/app/config.c
+++ b/app/config.c
@@ -100,7 +100,7 @@ bool            enable_bench       = true;
 bool            enable_mch_read    = true;
 bool            enable_numa        = false;
 
-bool            enable_ecc_polling = false;
+bool            enable_ecc_polling = true;
 
 bool            pause_at_start     = true;
 bool            dark_mode          = false;
diff --git a/system/imc/x86/amd_zen.c b/system/imc/x86/amd_zen.c
index 0cd7d7a..9e8328f 100644
--- a/system/imc/x86/amd_zen.c
+++ b/system/imc/x86/amd_zen.c
@@ -98,6 +98,7 @@ void get_imc_config_amd_zen(void)
         smn_reg = amd_smn_read(AMD_SMN_UMC_DRAM_ECC_CTRL + offset);
         if (smn_reg & (ECC_RD_EN | ECC_WR_EN)) {
             ecc_status.ecc_enabled = true;
+            imc.type = "DDR5-ECC";
 
             // Number of UMC to init
             uint8_t umc = 0, umc_max = 0;
```
(the second one is just to make sure the line `ecc_status.ecc_enabled = true` is executed, now I see it's redundant)

> Did you tried with latest nightly build? 

I assumed that nightly build is compiled with `enable_ecc_polling = false`, so there is no reason to try it.

**Reply by edo1** [reply]
tried on another cpu (9950x) with the same result: Linux and PassMark's MemTest86 detect ECC errors, but memtest86+ not, even if `enable_ecc_polling = true`<br>should an issue be opened?

**Reply by x86fr** [reply]
ECC polling for Granite Ridge CPUs are not supported yet. I'll check if I can add it before the public v8 release (end of September).

**Reply by edo1** [reply]
but I'm talking about AMD 7700 and 9950x CPUs<br>
spent about half an hour studying the code, it looks working, but for me it doesn't work

**Reply by debrouxl** [reply]
The Ryzen 9 9950X belongs to the family code named Granite Ridge, whereas the Ryzen 7 7700 is Raphael, as mentioned above by Sam.

**Reply by x86fr** [reply]
Raphael (7700) is supported in v8 (Both Vermeer and Raphael ECC code have been validated by AMD, so it should work). 
Granite Ridge (9950X) not yet.

**Reply by edo1** [reply]
> The Ryzen 9 9950X belongs to the family code named Granite Ridge,

Oops, my mistake. I thought AMD used one-word names, so I assumed it was an Intel codename. Thanks for the clarification.

> Raphael (7700) is supported in v8 (Both Vermeer and Raphael ECC code have been validated by AMD, so it should work)

But it doesn't for me.

**Reply by GrauBlitz** [reply]
Is there something like a table or list which CPUs are supported? I have some Socket 2011 Xeon CPUs with ECC memory but I do not think the support for them is already implemented?

EDIT: Looks like the only ECC check is indeed only for modern AMD Zen CPUs in [this line](https://github.com/memtest86plus/memtest86plus/blob/main/system/imc/x86/amd_zen.c#L95)
